src/type-data-frame.c

```diff
@@ -59,21 +59,48 @@ SEXP vctrs_new_data_frame(SEXP args) {
     Rf_errorcall(R_NilValue, "`x` must be a list");
   }
 
-  R_len_t size = df_size_from_list(x, n);
+  bool has_n = !Rf_isNull(n);
+  bool has_names = !Rf_isNull(r_names(x));
+
+  R_len_t size = -1;
+
+  SEXP out = PROTECT(r_maybe_duplicate(x));
+
+  for (SEXP attr_in = attrib; attr_in != R_NilValue; attr_in = CDR(attr_in)) {
+    SEXP tag = TAG(attr_in);
+
+    if (tag == R_NamesSymbol) {
+      // names are ignored if the input is named
+      if (has_names) continue;
+    } else if (tag == R_RowNamesSymbol) {
```

Comment:
_Style._ Shouldn't be an `else` branch here, since that suggests the previous branch hasn't `continue`'d.

Reply:
Disagree. The previous branch doesn't always continue. The `else`-s make it clear that only one of the alternatives is valid.

Reply:
You're right.